### ISSUE TITLE
[Material v5] Support mkdocs-localsearch plugin

### ISSUE THREAD

**Issue by wilhelmer** [squidfunk/mkdocs-material#1464]
_I checked that..._

- [x] ... the documentation does not mention anything about my idea
- [x] ... to my best knowledge, my idea wouldn't break something for other users
- [x] ... there are no open or closed issues that are related to my idea

## Description

As discussed in #1306 and mkdocs/mkdocs#1805, Material v5 breaks support for [mkdocs-localsearch](https://github.com/wilhelmer/mkdocs-localsearch).

To support the plugin and enable local search (file:// protocol), changes are required on your side. 

AFAIK, there are two major issues preventing local search using Material and current browsers:

1. **The search index is stored as a JSON, but loading JSONs isn't allowed on file protocol due to CORS.**<br>For Material v4, the [mkdocs-localsearch](https://github.com/wilhelmer/mkdocs-localsearch) plugin fixes this issue using a modified `fetch()` call. For Material v5, the same can be achieved, e.g., by modifying `ajax()` as outlined [here](https://github.com/squidfunk/mkdocs-material/issues/1306#issuecomment-571129344), but there's probably a better solution.
2. **The new v5 search engine uses a web worker, but in Chrome, web workers [aren't allowed](https://stackoverflow.com/questions/21408510/chrome-cant-load-web-worker) on file protocol.**<br>Since this is a new issue, this hasn't been addressed by mkdocs-localsearch before. A possible solution has been outlined [here](https://github.com/squidfunk/mkdocs-material/issues/1306#issuecomment-571323844) and [here](https://github.com/squidfunk/mkdocs-material/issues/1306#issuecomment-586928911).

I don't expect Material to support local search natively, but it would be great to be able to add that functionality using a plugin and without having to fork the entire Material project.

Any help is much appreciated.

### Use Cases

An MkDocs site is stored and accessed by users locally (file:// protocol), e.g., because the files are distributed with a desktop software. Users want to perform search on that site.

**Comment by squidfunk** [comment]
Thanks for creating this issue. As discussed, we can pass a `Promise` resolving with the contents of `search_index.json` to the initialization function, here:

https://github.com/squidfunk/mkdocs-material/blob/fdffefd53665a1d6a693be027df09d05e4c70574/src/base.html#L399-L406

The configuration will pass this through to the `setupSearchWorker` function, which is defined here:

https://github.com/squidfunk/mkdocs-material/blob/fdffefd53665a1d6a693be027df09d05e4c70574/src/assets/javascripts/workers/search/_/index.ts#L82-L128

As can be seen in line 106, the function will check, if a `Promise` was passed and otherwise issue the request itself. What remains is a patch for  `Worker`. I think it's better to patch `Worker` than to provide a fallback version of the search in the main thread, as the theme will probably make use of even more web workers in the future. See https://github.com/squidfunk/mkdocs-material/issues/1306#issuecomment-586928911 for an idea of a pseudo-worker implementation.

**Comment by wilhelmer** [comment]
I've been thinking about this issue a lot and spoken to other developers. In the end, I think it's not reasonable to continuously "re-patch" or "re-hack" a functionality that is actively blocked by numerous modern browsers. I'm sure we could work around the current restrictions, but with every new browser or site generator version, we may have to start all over again.

So I will stop working on mkdocs-localsearch. The plugin will not support Material v5.

In my documentation project, I will disable search in the local version and add a note that the search is only available online (including a link to the online version).

Thank you @squidfunk for your support, I'm sorry for the work you already put into this. Maybe someone else can pick this up if he/she desperately needs local search.

What you may consider though is adding some kind of visual hint that search doesn't work when the site is accessed via file:// protocol. Minimal solution: Implement a check for `location.protocol === "file:"` in JS and generate a console warning.

**Comment by squidfunk** [comment]
Let me think about this a bit. Sometimes the best solutions take time.

**Comment by squidfunk** [comment]
@wilhelmer just out of curiosity - what's the minimum browser versions that need to be supported? Can we ignore IE?

**Comment by squidfunk** [comment]
Okay, so I think I have a working polyfill for WebWorkers for the `file://` protocol. You may check out my new project [iframe-worker](https://github.com/squidfunk/iframe-worker). I just could not resist and had to see if it can be done. It even supports `importScripts`, error handling and is already tested in modern browsers. Some work needs to be done to support Edge and maybe even IE, which should be possible. It should be sufficient to add the following line to your `mkdocs.yml`:

``` yml
extra_javascript:
  - https://unpkg.com/iframe-worker
```

It's still to early too test it with the theme, but you could now adjust the local search plugin to extract the search index in whatever method you like into a `Promise` as described above.

**Comment by wilhelmer** [comment]
> @wilhelmer just out of curiosity - what's the minimum browser versions that need to be supported? Can we ignore IE?

Don't get me started on IE 😠 If I could drop support for it, I'd do it in a heartbeat, but it's still the default browser on all of our company's PCs. Also, according to Analytics, IE11 usage is >10% (!!!) among our customers. The enterprise business is a whole different ballgame. Crazy.

Anyway, regarding iframe-worker: Ohh, all that work for such an edge-case scenario! 😉 Just kidding, sounds great. If I have time, I'll try it this week.

**Comment by squidfunk** [comment]
> Don't get me started on IE 😠 If I could drop support for it, I'd do it in a heartbeat, but it's still the default browser on all of our company's PCs. Also, according to Analytics, IE11 usage is >10% (!!!) among our external visitors. The enterprise business is a whole different ballgame. Crazy.

The polyfill currently only supports browsers which provide a constructable `EventTarget` and `Promise` (which can both be polyfilled), but I didn't want to put the work into making it work in IE or Edge before considering this a viable approach.

> Anyway, regarding iframe-worker: Ohh, all that work for such an edge-case scenario! 😉 Just kidding, sounds great. If I have time, I'll try it this week.

It was fun to implement, as I haven't used the `postMessage` API for `iframe` communication before, but it was quite easy to implement. If you need any help (after reading https://github.com/squidfunk/mkdocs-material/issues/1464#issuecomment-586986321), I'd be happy to help.

**Comment by squidfunk** [comment]
... BTW, remember that Material v5 will drop support for Internet Explorer (all versions). IE is officially deprecated and supporting a browser that hasn't received any updates (except security patches) for the last 6 years which has <1% share among developers, the target group of this theme would hinder progress and innovation. We might add some logic to make it generally work, but I wouldn't want to maintain and backport fixes myself.

**Comment by wilhelmer** [comment]
Referring to issues 1) and 2) in my original post: Is 2) now solved automatically just by adding the polyfill to `extra_javascript`? And 1) has to be solved by providing an inlined search index using `Promise`?

**Comment by squidfunk** [comment]
1. Jup, correct. If you could prepare the work on the localsearch plugin (probably by just including the search index as a JavaScript file, and exposing it as a `Promise` in a global variable) it just needs to be passed to the theme. I'll wrap something up today how to pass it to the theme then.

2. I just saw that `extra_javascript` runs after the application, so it should be included before. What you can do is just extend the `scripts` block in a separate `main.html` template:

``` jinja
{% block scripts %}
  <script src="https://unpkg.com/iframe-worker/polyfill"></script>
  {{ block.super() }}
{% endbloc %}
```

**Comment by wilhelmer** [comment]
Solution for 2) seems to work fine, although I can't test it properly as long 1) isn't solved. (Couldn't find a Chrome setting that allows JSON file access, but still blocks web workers. The `--allow-file-access-from-files` flag allows both.)

Regarding 1), yes, it would be great if you could wrap something up. I don't fully understand it yet.

If I inline the search index as outlined above:

```html
<script id="__index" type="application/json">
  <!-- the contents of search_index.json -->
</script>
```

... then this has to be in a seperate file, right? I can't inline the 7 MB search index in every single page. But how would I embed that? Using an IFrame? 🤔 

So maybe, as you said in https://github.com/squidfunk/mkdocs-material/issues/1464#issuecomment-593867882, we should rather include the index as a JS file. I tried to expose it as a `Promise`, but that didn't work. Either I didn't do it correctly or the theme isn't ready for that yet?

**Comment by squidfunk** [comment]
> Either I didn't do it correctly or the theme isn't ready for that yet?

I'm working on it. Hope to finish it today. I'll tell you the exact place where the Promise needs to be passed so you can adjust the plugin then.

**Comment by squidfunk** [comment]
I had to do some restructuring today and haven't managed to finish the configurability of the search index, I'm sorry. However, regarding the implementation: in the end it doesn't matter how you provide the index, but it has to be wrapped in a `Promise`. This `Promise` must be passed to the theme at this location:

https://github.com/squidfunk/mkdocs-material/blob/f79075223f17368a9496b15291582f961811efd9/src/base.html#L397-L407

How? That I have to figure out tomorrow. I'll come back to you when I exposed the index in this configuration, but when you already have the index in a `Promise` it should work straight away.

**Comment by squidfunk** [comment]
> Solution for 2) seems to work fine, although I can't test it properly as long 1) isn't solved. (Couldn't find a Chrome setting that allows JSON file access, but still blocks web workers. The --allow-file-access-from-files flag allows both.)

`iframe-worker` will patch `Worker` always when served from `file://` protocol, so if search works when the flag is enabled it would mean that the worker polyfill works.

**Comment by wilhelmer** [comment]
> I'm sorry.

Omg no, take your time and don't be sorry for anything. You're doing exceptional work.

Side note: Integrating `iframe-worker` like this

```html
<script src="https://unpkg.com/iframe-worker/polyfill"></script>
```

kinda defeats the main purpose of serving docs via file:// protocol, which is making your docs available to people without an internet connection. So I guess most users should download the polyfill from `https://unpkg.com/iframe-worker@x.x.x/polyfill/index.js` and self-host that file in their project.

**Comment by squidfunk** [comment]
IMHO it does not defeat the purpose, because it depends on your aim, but if the distribution must be self-contained (including web fonts), you must include it, true. This can be easily done by downloading the polyfill and adding it to your `docs` directory so MkDocs will include it.

**Comment by squidfunk** [comment]
@FYI I tested Internet Explorer 11 out of curiosity and am quite astonished. Most of the business logic seems to work when I add some polyfills:

``` html 
{% block scripts %}
  <script src="https://unpkg.com/@babel/polyfill/dist/polyfill.min.js"></script>
  <script src="https://unpkg.com/url-polyfill/url-polyfill.min.js"></script>
  <script src="https://unpkg.com/mdn-polyfills/Element.prototype.closest"></script>
  <script src="https://unpkg.com/mdn-polyfills/Node.prototype.replaceWith"></script>
  <script src="https://unpkg.com/mdn-polyfills/Node.prototype.remove"></script>
  {{ super() }}
{% endblock %}
```

Then Material greats like this:

![Bildschirmfoto 2020-03-04 um 10 15 11](https://user-images.githubusercontent.com/932156/75863628-121ba000-5e01-11ea-89f8-0a50398ccdeb.png)

Off course there're no more colors, because hey, CSS variables, so one might need to define them oneself. Also, search does not work for some reason, but it should be doable with some more debugging. If you guys ant to go down that path, we can add a new section to the docs how to get Material working in IE again.

**Comment by wilhelmer** [comment]
I created a new `material-v5` branch in mkdocs-localsearch that you can use for testing.

`pip install git+https://github.com/wilhelmer/mkdocs-localsearch@material-v5`

That version should provide a Promise resolving with the contents of the search index:

![05-03-_2020_10-39-15](https://user-images.githubusercontent.com/2234833/75968780-3b563200-5ece-11ea-944f-17e54c8765c4.jpg)

**Comment by squidfunk** [comment]
Awesome, thanks!

**Comment by squidfunk** [comment]
A question: localsearch is a fork of search, right? Wouldn't it be easier for localsearch given the new way of configuration to co-exist with search, meaning `search` must be defined as a plugin and localsearch would just do two things.

1. Convert `search_index.json` to `search_index.js` and automatically (!) append it to the DOM in a pre-build step
2. Hand the `search_index` variable to the theme (in this case Material)
3. Automatically (download and) add the necessary polyfills

It would make things easier, as you wouldn't have to patch `header.html`. Or am I missing something?

**Comment by wilhelmer** [comment]
Yup, the plugin was more of a quick-and-dirty solution for myself that became popular against expectation. I have completely rewritten the plugin, it's no longer a fork of search. Instead, it only consists of simple file and DOM operations as outlined above. Feel free to try it, it's in the `material-v5` branch.

Note that `localsearch` must now be a separate entry in `mkdocs.yml`:

```yml
plugins:
    - localsearch
    - search
```

**Comment by squidfunk** [comment]
Perfect, then it should be even easier with v5. The latest HEAD contains the necessary code to make it work with:

``` jinja
{% extends "base.html" %}

{% block config %}
  <script src="https://unpkg.com/iframe-worker/polyfill"></script>
  <script>
    var __config = {
      search: {
        index: Promise.resolve(index)
      }
    }
  </script>
{% endblock %}
```

**Comment by squidfunk** [comment]
... you could download and inline `iframe-worker` as part of the plugin, but I think for general testing we could first assume an Internet connection. If it works, we can take it from here. Note however that the configuration structure might change before release again, but that should only be a cosmetic change.

**Comment by wilhelmer** [comment]
Great, I'll try that tomorrow.

Regarding `iframe-worker`: I think it's not too much to ask from users to extend the `scripts` block with the polyfill [as you described](https://github.com/squidfunk/mkdocs-material/issues/1464#issuecomment-593867882). There, they can either specifiy the http version or a self-hosted file.

**Comment by squidfunk** [comment]
Makes sense, as versioning both packages separately is definitely better. I'll release a new beta by tomorrow, so maybe you can just try that.

**Comment by wilhelmer** [comment]
I tested it with Beta 3.

The good news: No more console errors (FF + Chrome)
The bad news: Search doesn't work, no search results displayed.

Here's a test project:
[test-localsearch.zip](https://github.com/squidfunk/mkdocs-material/files/4297577/test-localsearch.zip)

**Note:** Make sure to update to the latest version of mkdocs-localsearch (`material-v5` branch). I had to remove the adding of `search_index.js` to `extra_javascript`, because the script must be referenced **before** the theme scripts, preferably in the `config` block:

```html
{% block config %}
<script src="https://unpkg.com/iframe-worker/polyfill"></script>
<script src="search/search_index.js"></script>
<script>
    var __config = {
        search: {
            index: Promise.resolve(index)
        }
    }
</script>
{% endblock %}
```

**Comment by squidfunk** [comment]
@wilhelmer I debugged it, and it's some kind of race condition. When I delay the setup of search by `1ms`, it works. I'll see if we need to patch `iframe-worker` or Material.

**Comment by wilhelmer** [comment]
Okay, thanks.
I noticed that `index` is resolved twice, in `__config` and in the JS file. That's redudant, right? We only need one `Promise` or the other?

**Comment by squidfunk** [comment]
That right, but it shouldn't matter, as the `Promise` constructor will just return the value you pass to it by reference if it's already `Promise`. It's actually _the_ way of checking for `Promise` support.

**Comment by squidfunk** [comment]
Should be fixed with 3a25eb46d4e42fcd91f0cfbfe70b50818e6b1c2f. It ensures that index initialization is done on the next tick when all event handlers are set up. My research suggests that it has nothing to do with `iframe-worker`, but with the fact that when the search index is loaded with XHR it takes some time, as opposed to the `Promise` which resolves instantly. I see if we can improve on this situation, but for now it should be sufficient.

**Comment by squidfunk** [comment]
I'll release a new beta when we collected some more feedback, but you may try the `HEAD`.

**Comment by wilhelmer** [comment]
Works in Firefox, doesn't work in Chrome and Edge Chromium.
With the monkey patch, it works in all browsers.

**Comment by squidfunk** [comment]
I tested in Chrome, actually. However, as noted, I have to check whether we can improve on the current solution. The important part is: it works!

**Comment by wilhelmer** [comment]
Okay, the above is true, but only when using my very large 23 MB prebuilt search index.
With smaller indexes, it works fine in all browsers.

**Comment by squidfunk** [comment]
Okay, cool, so we have a POC now. Pretty awesome. I'll check how we can improve it, but I'll need to invest some time in other parts of the theme first.

**Comment by wilhelmer** [comment]
Yes. I hereby solemnly declare that through the tireless work of @squidfunk, mkdocs-localsearch has risen from the dead and we have a working version for v5! 🎉

**Comment by squidfunk** [comment]
Great, so this issue can be closed? It's already supported in v4 and now that we have support in v5 we can close it. We can open a new issue after the release of v5 to see if and how we can improve it.

**Comment by wilhelmer** [comment]
![makeitso](https://user-images.githubusercontent.com/2234833/76083553-323c9200-5fae-11ea-881d-41ffd2aca2e9.jpeg)
